runcap.man
.TH runcap 8
'''
.SH NAME
\fBruncap\fR \- run command with specified permissions and/or capabilities.
'''
.SH SYNOPSIS
\fBruncap\fR [\fIuser\fR][\fB:\fIgroup\fR]... \fIcommand\fR \fIargs\fR ...
'''
.SH DESCRIPTION
\fBruncap\fR runs specified command, possibly setting effective uid/gid,
ulimits, capabilities and cgroups.
.P
This tool is intended for running daemons with non-root effective
uids while still allowing them to do their privileged tasks.
.P
\fBWARNING\fR: due to weird capability inheritance rules in non-patched
Linux kernels, setting capabilities will not work as expected right away.
See POSIX 1003.1e notes below.
'''
.SH USAGE
Permissions specification is a single argument and should be supplied without
any spaces. All parts are optional, however in most cases \fIuser\fR should
be specified.
.IP "\fIuser\fR" 4
User, see \fBsetuid\fR(2).
.IP "\fB:\fIgroup\fR" 4
Primary group, for \fBsetgid\fR(2) call.
.IP "\fB@\fIgroup\fB,\fIgroup\fB,\fI...\fR" 4
Supplementary groups, see \fBsetgroups\fR(2).
.IP "\fB.\fR[\fIctdfxlqenpmETis\fR]\fIN\fR[\fB/\fIM\fR][\fIkMGmh\fR]" 4
Process limits, see below.
.IP "\fB+\fIcap\fR" 4
Retain named capability, see \fBcapabilities\fR(7) and Capabilities below.
.P
Several \fB.\fR or \fB+\fR blocks can be specified. Users and groups can be specified
as names or as (numeric) ids. When names are used, /etc/passwd and /etc/group will be
checked to get relevant ids.
.br
Example: "user:group@sg1,sg2.n10.c0+bind+nice".
.P
Process limits with respective C constants (see \fBsetrlimits\fR(2)):
.PD 0
.RS
.TP
\fBa\fR size of the process's virtual memory (RLIMIT_AS)
.TP
\fBc\fR size of core files created (RLIMIT_CORE)
.TP
\fBt\fR amount of cpu time in seconds (RLIMIT_CPU)
.TP
\fBd\fR size of a process's data segment (RLIMIT_DATA)
.TP
\fBf\fR size of files written (RLIMIT_FSIZE)
.TP
\fBx\fR number of file locks (RLIMIT_LOCKS)
.TP
\fBl\fR size that may be locked into memory (RLIMIT_MEMLOCK)
.TP
\fBq\fR number of bytes in POSIX message queues (RLIMIT_MSGQUEUE)
.TP
\fBe\fR scheduling priority (RLIMIT_NICE)
.TP
\fBn\fR number of open files (RLIMIT_NOFILE)
.TP
\fBp\fR number of threads (RLIMIT_NPROC)
.TP
\fBm\fR resident set size (RLIMIT_RSS)
.TP
\fBE\fR real-time scheduling priority (RLIMIT_RTPRIO)
.TP
\fBT\fR time in microseconds process with real-time
schedule policy may consume without making a blocking syscall (RLIMIT_RTTIME)
.TP
\fBi\fR number of pending signals (RLIMIT_SIGPENDING)
.TP
\fBs\fR stack size (RLIMIT_STACK)
.RE
.P
Full list of capabilities, with respective C constants:
.PD 0
.RS
'''
.TP
\fBchown\fR (CAP_CHOWN)
.TP
\fBdac-override\fR (CAP_DAC_OVERRIDE)
.TP
\fBdac-read-search\fR (CAP_DAC_READ_SEARCH)
.TP
\fBfowner\fR (CAP_FOWNER)
.TP
\fBfsetid\fR (CAP_FSETID)
.TP
\fBkill\fR (CAP_KILL)
.TP
\fBsetgid\fR (CAP_SETGID)
.TP
\fBsetuid\fR (CAP_SETUID)
.TP
\fBsetpcap\fR (CAP_SETPCAP)
.TP
\fBimmutable\fR (CAP_LINUX_IMMUTABLE)
.TP
\fBbind\fR (CAP_NET_BIND_SERVICE)
.TP
\fBbroadcast\fR (CAP_NET_BROADCAST)
.TP
\fBnetadmin\fR (CAP_NET_ADMIN)
.TP
\fBnetraw\fR (CAP_NET_RAW)
.TP
\fBipclock\fR (CAP_IPC_LOCK)
.TP
\fBipcowner\fR (CAP_IPC_OWNER)
.TP
\fBmodule\fR (CAP_SYS_MODULE)
.TP
\fBrawio\fR (CAP_SYS_RAWIO)
.TP
\fBchroot\fR (CAP_SYS_CHROOT)
.TP
\fBptrace\fR (CAP_SYS_PTRACE)
.TP
\fBpacct\fR (CAP_SYS_PACCT)
.TP
\fBsysadmin\fR (CAP_SYS_ADMIN)
.TP
\fBboot\fR (CAP_SYS_BOOT)
.TP
\fBnice\fR (CAP_SYS_NICE)
.TP
\fBresource\fR (CAP_SYS_RESOURCE)
.TP
\fBtime\fR (CAP_SYS_TIME)
.TP
\fBttyconfig\fR (CAP_SYS_TTY_CONFIG)
.TP
\fBmknod\fR (CAP_MKNOD)
.TP
\fBlease\fR (CAP_LEASE)
.TP
\fBauditrewrite\fR (CAP_AUDIT_WRITE)
.TP
\fBauditcontrol\fR (CAP_AUDIT_CONTROL)
.TP
\fBsetfcap\fR (CAP_SETFCAP)
.TP
\fBmacoverride\fR (CAP_MAC_OVERRIDE)
.TP
\fBmacadmin\fR (CAP_MAC_ADMIN)
.TP
\fBsyslog\fR (CAP_SYSLOG)
.TP
\fBwakealarm\fR (CAP_WAKE_ALARM)
.TP
\fBblocksuspend\fR (CAP_BLOCK_SUSPEND)
'''
.RE
.P
Refer to \fBcapabilities\fR(7) for detailed description.
.PD 1
.P
'''
.SS POSIX 1003.1e notes
Mainline Linux kernels implement draft POSIX 1003.1e notion of capabilities.
POSIX capabilities are tied to applications (executable files), not to users
and/or running processes. This is at odds with common Unix uid/gid permissions and ACLs.
.P
POSIX capabilities provide no way for a parent process to force permissions onto
an unsuspecting child. Capabilities expected for a particular executable
must be set explicitly as extended file attributes using \fBsetxattr\fR.
.P
Make sure to check \fBcapabilities\fR(7) before trying to make it work, paying especial
attention to "File capabilities" and "Transformation of capabilities during execve".
.P
runcap can only set \fIinheritable\fR process capabilities.
For those capabilities to actually take effect, the file being executed
must have relevant bits set in \fIinheritable\fR part of its security.capabilities
extended attribute.
'''
.SH ENVIRONMENT
$PATH is used to find specified command. See \fBexecvp\fR(1).
'''
.SH NOTES
runcap is in many ways similar to \fBsu\fR(1) when su is used by root
to invoke non-privileged user. Crucial difference however is that runcap
allows retaining some superuser capabilities across setuid/setgid.
.P
runcap can also be viewed as the opposite of \fBsudo\fR(1).
sudo is expected to be run by non-privileged user, selectively and temporarily
granting additional permissions. In contrast, runcap is expected
to be used by root to lower permissions selectively.
.P
runcap should never be installed suid root. In fact, since non-privileged users
have no capabilities to drop and can not change their effective uids
at will, runcap should not be made available to non-privileged users at all.
'''
.SH SEE ALSO
Similar utilities: \fBsu\fR(1), \fBsudo\fR(1), \fBrunuser\fR(1), \fBulimit\fR in \fBbash\fR(1),
\fBnice\fR(1), \fBcgexec\fR(1), \fBcapsh\fR(1).
.P
Syscalls: \fBsetresuid\fR(2), \fBsetresgid\fR(2), \fBsetrlimit\fR(2),
\fBsetpriority\fR(2), \fBcapset\fR(2), \fBexecve\fR(2).
.P
General concepts: \fBcredentials\fR(7), \fBcapabilities\fR(7),
.br
Documentation/cgroups/ from kernel source tree.
.P
File attributes: \fBattr\fR(5), \fBsetcap\fR(8), \fBsetxattr\fR(2).


runcap.c
/* runcap is a companion utility to sninit, in the sense it does what
   could have been implemented in sninit (but wasn't).
   See doc/limits.txt for discussion of the problem.

   Yet even with background like this, runcap is completely independent
   from sninit. It could easily be packaged and distributed as a stanalone utility.

   This perfect separation of functions was actually one of the main reasons
   to make it standalone, vs. having all this built into init. */

#define _GNU_SOURCE
#include <stdlib.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdint.h>
#include <sys/stat.h>
#include <sys/mman.h>
#include <sys/resource.h>
#include <sys/prctl.h>
#include <string.h>
#include <errno.h>

/* This should have been in sys/capability.h, but that header
   is non-standard, and hosts pretty ugly libcap stuff aside
   of capget/capset syscalls. */
#include "sys_capability.h"
extern int capset(cap_user_header_t hdrp, cap_user_data_t datap);

/* Size limit for mmaped /etc/passwd and /etc/group */
#define MAXFILELEN 65536
/* Maximum number of secondary groups */
#define MAXGROUPS 20

#define ERRNO ((void*)-1)

/* Now technically this is incorrect, cgroups can be mounted at any location.
   libcgroup uses /etc/cgconfig.conf to find the path, a C-syntax file with
   nested blocks for no apparent reason. Parsing that in a small utility does
   not look like a good idea.
   However, for a given system, it's probably safe to assume the path is fixed. */
#define CGBASE "/sys/fs/cgroup"

/* Stored process limitations: uid/gids and capabilities */
/* Cgroups and ulimits are applied immediately, no need to store those */
int uid = -1;
int gid = -1;
/* Secondary groups */
int gidn = 0;
gid_t gids[MAXGROUPS];
/* Capabilities */
uint32_t caps[VFS_CAP_U32] = { 0 };

/* Files for looking up uids and gids by name */
struct rcfile {
	const char* name;
	char* buf;
	size_t len;
	char* ls;		/* line start */
	char* le;		/* line end */
} passwd = {
	.name = "/etc/passwd"
}, groups = {
	.name = "/etc/group"
};

/* ulimit keys, closely following bash(1) ulimit command */
/* note T is used for RTTIME here; in bash, it's apparently NPROC */
struct limitdesc {
	char key;
	int resource;
} ulimits[] = {
	{ 'a', RLIMIT_AS },
	{ 'c', RLIMIT_CORE },
	{ 't', RLIMIT_CPU },
	{ 'd', RLIMIT_DATA },
	{ 'f', RLIMIT_FSIZE },
	{ 'x', RLIMIT_LOCKS },
	{ 'l', RLIMIT_MEMLOCK },
	{ 'q', RLIMIT_MSGQUEUE },
	{ 'e', RLIMIT_NICE },
	{ 'n', RLIMIT_NOFILE },
	{ 'p', RLIMIT_NPROC },
	{ 'm', RLIMIT_RSS },
	{ 'E', RLIMIT_RTPRIO },
	{ 'i', RLIMIT_SIGPENDING },
	{ 's', RLIMIT_STACK },
#ifdef RLIMIT_RTTIME
	{ 'T', RLIMIT_RTTIME },		/* musl lacks this? */
#endif
	{ '\0' }
};

/* Using shorter names for capabilities here, things like cap_net_bind_service
   are way too long to type and make no sense anyway. */
struct capdesc {
	const char* key;
	int cap;
} capabilities[] = {
	{ "chown", 		CAP_CHOWN },
	{ "dac-override", 	CAP_DAC_OVERRIDE },
	{ "dac-read-search",	CAP_DAC_READ_SEARCH },
	{ "fowner",		CAP_FOWNER },
	{ "fsetid",		CAP_FSETID },
	{ "kill",		CAP_KILL },
	{ "setgid",		CAP_SETGID },
	{ "setuid",		CAP_SETUID },
	{ "setpcap",		CAP_SETPCAP },
	{ "immutable",		CAP_LINUX_IMMUTABLE },
	{ "bind",		CAP_NET_BIND_SERVICE },
	{ "broadcast",		CAP_NET_BROADCAST },
	{ "netadmin",		CAP_NET_ADMIN },
	{ "netraw",		CAP_NET_RAW },
	{ "ipclock",		CAP_IPC_LOCK },
	{ "ipcowner",		CAP_IPC_OWNER },
	{ "module",		CAP_SYS_MODULE },
	{ "rawio",		CAP_SYS_RAWIO },
	{ "chroot",		CAP_SYS_CHROOT },
	{ "ptrace",		CAP_SYS_PTRACE },
	{ "pacct",		CAP_SYS_PACCT },
	{ "sysadmin",		CAP_SYS_ADMIN },
	{ "boot",		CAP_SYS_BOOT },
	{ "nice",		CAP_SYS_NICE },
	{ "resource",		CAP_SYS_RESOURCE },
	{ "time",		CAP_SYS_TIME },
	{ "ttyconfig",		CAP_SYS_TTY_CONFIG },
	{ "mknod",		CAP_MKNOD },
	{ "lease",		CAP_LEASE },
	{ "auditrewrite",	CAP_AUDIT_WRITE },
	{ "auditcontrol",	CAP_AUDIT_CONTROL },
	{ "setfcap",		CAP_SETFCAP },
	{ "macoverride",	CAP_MAC_OVERRIDE },
	{ "macadmin",		CAP_MAC_ADMIN },
	{ "syslog",		CAP_SYSLOG },
	{ "wakealarm",		CAP_WAKE_ALARM },
	{ "blocksuspend",	CAP_BLOCK_SUSPEND },
	{ NULL }
};

static void die(const char* msg, const char* arg, const char* err) __attribute__((noreturn));

static void parsecaps(char* caps);

static void adduser(char* p);
static void addgroup(char* p);
static void addsgroup(char* p);
static void setlimit(char* p);
static void setcg(char* p);
static void addcap(char* p);

static void applycaps();

int main(int argc, char** argv)
{
	if(argc < 3) 
		die("Usage: runcap [capabilities] command arg arg arg ...", NULL, NULL);

	parsecaps(argv[1]);
	applycaps();

	execvp(argv[2], argv + 2);
	die("Can't exec ", argv[2], ERRNO);
}

/* user:group@subgroup1,subgroup2.n100s65M.T10+bind */
static void parsecaps(char* caps)
{
	char* p = caps;
	char* q;
	char qs;
	char mode = '~';

	while(*p) {
		switch(*p) {
			case ':':
			case '@':
			case '.':
			case '+': 
			case '%': mode = *p;
			case ',': p++;
		}
		for(q = p; *q && !strchr(":%@.+,", *q); q++);
		/* user@group,subgroup1... */
		/* p ---^    ^--- q        */
		qs = *q; *q = '\0';
		/* user@group₀subgroup1... */
		switch(mode) {
			case '~': adduser(p); break;
			case ':': addgroup(p); break;
			case '@': addsgroup(p); break;
			case '+': addcap(p); break;
			case '.': setlimit(p); break;
			case '%': setcg(p); break;
		};
		*q = qs; p = q;
		/* user@group,subgroup1... */
		/*           ^--- p        */
	};
}

/* rc* to avoid name clashes with init_conf_mem, keeping ctags consistent */
/* Note there's no munmap; mmaped files are dropped during exec, so why bother */
static void rcmapfile(struct rcfile* f)
{
	int fd;
	struct stat st;

	if((fd = open(f->name, O_RDONLY)) < 0)
		die("Can't open ", f->name, ERRNO);
	if(fstat(fd, &st))
		die("Can't stat ", f->name, ERRNO);
	if(st.st_size > MAXFILELEN)
		die("Can't mmap ", f->name, "the file is too large");
	if((f->buf = mmap(NULL, st.st_size, PROT_READ | PROT_WRITE, MAP_PRIVATE, fd, 0)) == MAP_FAILED)
		die("Can't mmap ", f->name, ERRNO);

	close(fd);
	f->len = st.st_size;

	f->ls = NULL;
	f->le = NULL;
};

static void rcrewind(struct rcfile* f)
{
	f->ls = NULL;
	f->le = NULL;
};

static char* rcnextline(struct rcfile* f)
{
	if(f->le)
		*(f->le) = '\n';
	f->ls = (f->le ? f->le + 1 : f->buf);
	if(f->ls >= f->buf + f->len)
		return NULL;
	
	for(f->le = f->ls; *(f->le) && *(f->le) != '\n'; f->le++);
	*(f->le) = '\0';

	return f->ls;
};

static int isnumeric(char* s)
{
	char* p;
	for(p = s; *p; p++)
		if(*p < '0' || *p > '9')
			return 0;
	return 1;
}

static void adduser(char* user)
{
	char* l;
	char* name;
	char* uidstr;
	char* gidstr;

	if(isnumeric(user)) {
		uid = atoi(user);
		return;
	}

	rcmapfile(&passwd);
	while((l = rcnextline(&passwd))) {
		name = strsep(&l, ":");
		if(strcmp(name, user))
			continue;
		strsep(&l, ":");
		uidstr = strsep(&l, ":");
		gidstr = strsep(&l, ":");
		if(!uidstr || !gidstr)
			continue;
		uid = atoi(uidstr);
		gid = atoi(gidstr);
		return;
	}

	die("Unknown user ", user, NULL);
};

static gid_t findgroup(char* group)
{
	char* l;
	char* grpname;
	char* gidstr;

	if(isnumeric(group))
		return atoi(group);

	if(!groups.buf)
		rcmapfile(&groups);
	else
		rcrewind(&groups);

	while((l = rcnextline(&groups))) {
		grpname = strsep(&l, ":");
		if(strcmp(grpname, group))
			continue;
		strsep(&l, ":");
		gidstr = strsep(&l, ":");
		return atoi(gidstr);
	};

	die("Unknown group ", group, NULL);
}

static void addgroup(char* group)
{
	gid = findgroup(group);
};

static void addsgroup(char* group)
{
	if(gidn >= MAXGROUPS)
		die("Too many groups", NULL, NULL);
	gids[gidn++] = findgroup(group);
};

/* Limits could have been stored just like uid/gids and caps... except it's not needed.
   Unlike uid/gid/cap changes which must be coordinated, limit settings are independent
   and do not prevent runcap from doing its work even if set immediately. So, why bother. */
static void setlimit(char* lim)
{
	struct limitdesc* p;
	char* hard;
	char* soft;
	struct rlimit rl;

	for(p = ulimits; p->key; p++)
		if(p->key == *lim)
			break;
	if(!p->key)
		die("Unknown limit key ", lim, NULL);

	char* q = lim + 1;
	soft = strsep(&q, "/");
	hard = q;

	rl.rlim_cur = atoi(soft);
	rl.rlim_max = hard ? atoi(hard) : rl.rlim_cur;

	if(setrlimit(p->resource, &rl))
		die("Can't set ulimit ", lim, ERRNO);
};

/* may be used by strerror, so exported */
char* ltoa(long num)
{
	static char buf[16];
	int i = sizeof(buf)-1;
	buf[i--] = '\0';
	while(num && i >= 0) {
		buf[i--] = num % 10;
		num /= 10;
	};
	return buf + i + 1;
}

static void setcg(char* cg)
{
	int cgl = strlen(CGBASE) + strlen(cg) + 15;
	char cgp[cgl];
	int fd;
	char* pid;

	strcpy(cgp, CGBASE);
	strcat(cgp, "/");
	strcat(cgp, cg);
	strcat(cgp, "/tasks");

	if((fd = open(cgp, O_WRONLY)) < 0)
		die("Can't open ", cgp, ERRNO);

	pid = ltoa(getpid());
	if(write(fd, pid, strlen(pid)) <= 0)
		die("Can't add process to ", cgp, ERRNO);
	close(fd);
}

static void addcap(char* cap)
{
	struct capdesc* p;

	for(p = capabilities; p->key; p++)
		if(!strcmp(p->key, cap))
			break;
	if(!p->key)
		die("Unknown capability ", cap, NULL);

	caps[CAP_TO_INDEX(p->cap)] |= CAP_TO_MASK(p->cap);
};

/* Ulimits, if any, have been set earlier. That doesn't affect anything here. */
static void applycaps()
{
	struct __user_cap_header_struct caphdr = {
		.version = _LINUX_CAPABILITY_VERSION_3,
		.pid = 0
	};
	struct __user_cap_data_struct capepi[VFS_CAP_U32];
	int hascaps = 0;
	int i;

	for(i = 0; i < VFS_CAP_U32; i++) {
		capepi[i].effective = caps[i];
		capepi[i].permitted = caps[i];
		capepi[i].inheritable = caps[i];
		if(caps[i]) hascaps = 1;
	}
	
	if(hascaps)
		if(prctl(PR_SET_KEEPCAPS, 1L))
			die("prctl failed", NULL, ERRNO);
	if(gid >= 0)
		if(setresgid(gid, gid, gid))
			die("setresgid failed", NULL, ERRNO);
	if(uid >= 0)
		if(setresuid(uid, uid, uid))
			die("setresuid failed", NULL, ERRNO);
	if(hascaps)
		if(capset(&caphdr, capepi))
			die("capset failed", NULL, ERRNO);
};

static char* strlncat(char* buf, char* ptr, int buflen, const char* str)
{
	int len = strlen(str);
	int avail = (buf + buflen) - ptr;
	if(len > avail) len = avail;
	memcpy(ptr, str, len);
	return ptr + len;
}

static void die(const char* msg, const char* arg, const char* err)
{
	int buflen = 1024;
	char buf[buflen+1];
	char* ptr = buf;

	ptr = strlncat(buf, ptr, buflen, msg);
	if(arg)
		ptr = strlncat(buf, ptr, buflen, arg);
	if(err == ERRNO)
		err = strerror(errno);
	if(err) {
		ptr = strlncat(buf, ptr, buflen, ": ");
		ptr = strlncat(buf, ptr, buflen, err);
	}
	*ptr = '\n';

	write(2, buf, ptr - buf + 1);
	_exit(-1);
}
